cppabi32/cppabi32.rst

```diff
@@ -0,0 +1,1698 @@
+    }  
+    extern "C" void __aeabi_vec_delete3(
+            void* user_array, void* (*destructor)(void*), void (*dealloc)(void*, size_t))
+    { // The meaning of this function is given by the following model implementation...
+      // like:  __cxa_vec_delete3(user_array, element_size_of(user_array),
+      //                          cookie_size, destructor, decalloc);
+      if (user_array != NULL) {
+          size_t size =
+              element_size_of(user_array) * element_count_of(user_array) + cookie_size;
+        void *array_cookie;
+        try {
+          array_cookie = __aeabi_vec_dtor_cookie(user_array, destructor);
+        } catch (...) {
+          try {
+            (*dealloc)(&cookie_of(user_array), size);
+          } catch (...) {
+             std::terminate();
+          }
+          throw;
+        }
+        (*dealloc)(array_cookie, size);
+      }
+    }  
+    extern "C" void __aeabi_vec_delete3_nodtor(
+            void* user_array, void (*dealloc)(void*, size_t))
+    { // The meaning of this function is given by the following model implementation...
+      // like:  __cxa_vec_delete3(user_array, element_size_of(user_array),
+      //                          cookie_size, 0, decalloc);
+      if (user_array != NULL) {
+          size_t size =
+              element_size_of(user_array) * element_count_of(user_array) + cookie_size;
+        (*dealloc)(&cookie_of(user_array), size);
+      }
+    }
+    extern "C" int  __aeabi_atexit(void* object, void (*destroyer)(void*), void* dso_handle)
+    { // atexit(f) should call __aeabi_atexit (NULL, f, NULL)
+      // The meaning of this function is given by the following model implementation... 
+      return __cxa_atexit(destroyer, object, dso_handle);           // 0 ==> OK; non-0 ==> failed
+    }
+  } // namespace __aeabiv1
+
+Code examples for the delete expression
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Section 5.3.5 of the ISO C++ standard discusses the delete expression.
+
+The code needed to implement ``delete [] p`` is tabulated in `Implementation of
+delete [] p`_, below. It depends on:
+
+-  The static element type of p (referred to as T below),
+
+-  Which ``operator delete []`` is being used for this de-allocation: either
+   ``::operator delete(void*)`` or ``T1::operator delete(void*)``, where T1 is
+   T or a base class of T.
+
+-  Whether a cookie is needed for arrays of T (see `Array cookies`_).
+
+-  *Has dtor*, which means T is a class type with a non-trivial
+   destructor [ISO C++ §12.4]. In cases where there is no cookie there
+   must be no dtor.
+
+.. _Implementation of delete [] p:
+
+.. table:: Implementation of delete [] p
+
+  +-------------------------------+--------------------+------------------+------------------------------------------------------------------+------------+
+  | ``operator delete []``        | Needs cookie       | *Has dtor*       | Implementation of ``delete [] p / ::delete [] p``                | Note       |
```

Comment:
a bit of a nit, but perhaps in the first header column, perhaps we could get the `[]` of delete on the same line?

another nit: in the last three rows of the 4th column, the expression breaks across the line in unfortunate ways